src/data-model-providers/codegen/CodegenDataModelProvider_Read.cpp

```diff
@@ -104,6 +104,11 @@ DataModel::ActionReturnStatus CodegenDataModelProvider::ReadAttribute(const Data
 
     if (auto * cluster = mRegistry.Get(request.path); cluster != nullptr)
     {
+        // Attribute must be readable to allow a read request for it.
+        // Subscriptions and read request may send invalid values to paths and Attribute path expansion. Such validations
+        // are not done by callers however ServerClusterInterface API requirement is to ensure path validity
```

Comment:
Why is this validation not done generically?  It's part of the IM spec, so it has to happen, and has to happen in the right order with other checks, so ideally it would not be up to every provider to handle it....

In particular, the way the spec is being written for this the order of checks will need to be:

1. ACL check.
2. Existence check.
3. Readability check.
4. Another ACL check.

That will all need to happen in RetrieveClusterData in reporting engine, presumably, right?

Reply:
Apparently it is not done so. 

https://github.com/project-chip/connectedhomeip/blob/master/src/data-model-providers/codegen/CodegenDataModelProvider_Read.cpp#L110 has `FindAttributeMetadata` today and validates existence and returns appropriate statuses. 

I noticed this when I converted code from ember to codegen: ember was doing checks, codegen was not in the same piece of code.

This applies to write as well.

Reply:
I updated the logic. Blast radius increased significantly.